components/builder-api-client/src/builder.rs

```diff
@@ -1426,6 +1430,45 @@ impl BuilderAPIClient {
     }
 }
 
+/// Retry API operations within reasonable bounds.
+/// Provide an Option with a custom back-off algorithm from the `retry` crate
+/// or a default Fibonacci algorithm will be used. Unrecoverable HTTP errors
+/// will not be retried.
+#[macro_export]
+macro_rules! retry_builder_api {
+    ($iter_opt:expr, $api_future:expr) => {
```

Comment:
Declarative macros like this can match multiple different cases. As such, you actually don't have to invoke this with an explicit `None` in the first position if you want to use the Fibonacci delay strategy; you could just have that be another match case of the macro.

So, to use the default strategy, you could call it like:
```rust
retry_builder_api!(future);
```
and to call it with the custom iterator, you'd use

```rust
retry_builder_api!(iterator, future)
```
If you want to get fancier, you could tweak the macro for readability, so you could call it as something like:
```rust
retry_builder_api!(future, with_custom_iterator: iterator);
```
(As currently coded, I don't think this has to be a macro, but it's probably a relatively cheap way to get optional parameters as described above without having to create some new types, so that's fine.)

Reply:
> As currently coded, I don't think this has to be a macro

I believe `retry_builder_api` has to be a macro because futures can only be `await`ed once. We dont want to resolve the expression that was passed into `retry_builder_api`. That would only give is the ability to try the future once. Instead, we want to preserve the expression used to generate the future so we can call that expression multiple times.

This all stems from a lack of true async closures. If we had async closures we could pass a closure to generate the future instead of resorting to a macro.

Reply:
@davidMcneil is correct, this is the exact reason I had to go down this macro path.

Reply:
Ah, cool :+1:

Reply:
I like @christophermaier suggestion [here](https://github.com/habitat-sh/habitat/pull/7874#discussion_r470732256) of allowing the `retry_builder_api` macro to be called with or without a retry iterator. I think it would look something like:

```rust
#[macro_export]
macro_rules! retry_builder_api {
    ($api_future:expr) => {
        retry_builder_api!($api_future,
                           retry::delay::Fibonacci::from(API_RETRY_WAIT).take(API_RETRIES))
    };
    ($api_future:expr,iter_opt: expr,) => {
        async {
            ...
        }
    };
}
```

Or we could simply hardcode the retry iterator. It looks like we only ever call the macro with `None`.

Reply:
I love this idea @christophermaier @davidMcneil!
I've refactored the macro in https://github.com/habitat-sh/habitat/pull/7874/commits/26704a92a178597076d034d64aa498acfd33132d allowing for it to be called with or without the iterator and including a `with_custom_iterator: ` for readability.